control-plane-pki-operator/certificatesigningcontroller/certificatesigningcontroller.go

```diff
@@ -0,0 +1,173 @@
+	}
+	if requeue {
+		return factory.SyntheticRequeueError
+	}
+	if validationErr != nil {
+		syncContext.Recorder().Eventf("CertificateSigningRequestInvalid", "%q is invalid: %s", name, validationErr.Error())
+	} else {
+		syncContext.Recorder().Eventf("CertificateSigningRequestValid", "%q is valid", name)
+	}
+	if cfg != nil {
+		_, err := c.kubeClient.CertificatesV1().CertificateSigningRequests().ApplyStatus(ctx, cfg, metav1.ApplyOptions{FieldManager: c.fieldManager})
+		if err != nil && validationErr != nil {
+			syncContext.Recorder().Eventf("CertificateSigningRequestFulfilled", "%q in %q is fulfilled", name)
+		}
+		return err
+	}
+
+	return nil
+}
+
+const backdate = 5 * time.Minute
+
+func (c *CertificateSigningController) processCertificateSigningRequest(name string, now func() time.Time) (*certificatesv1applyconfigurations.CertificateSigningRequestApplyConfiguration, bool, error, error) {
+	csr, err := c.getCSR(name)
+	if err != nil {
+		return nil, false, nil, err
+	}
+
+	// Ignore unapproved or failed requests
+	if !certificates.IsCertificateRequestApproved(csr) || certificates.HasTrueCondition(csr, certificatesv1.CertificateFailed) {
+		return nil, false, nil, nil
+	}
+
+	// Ignore signed requests
+	if csr.Status.Certificate != nil {
+		return nil, false, nil, nil
+	}
+
+	x509cr, err := certificates.ParseCSR(csr.Spec.Request)
+	if err != nil {
+		return nil, false, nil, fmt.Errorf("unable to parse csr %q: %v", csr.Name, err)
+	}
+	if validationErr := c.validator(csr, x509cr); validationErr != nil {
+		cfg := certificatesv1applyconfigurations.CertificateSigningRequest(name)
+		cfg.Status = certificatesv1applyconfigurations.CertificateSigningRequestStatus().WithConditions(
+			certificatesv1applyconfigurations.CertificateSigningRequestCondition().
+				WithType(certificatesv1.CertificateFailed).
+				WithStatus(corev1.ConditionTrue).
+				WithReason("SignerValidationFailure").
+				WithMessage(validationErr.Error()).
+				WithLastUpdateTime(metav1.Now()),
+		)
+		return cfg, false, validationErr, nil
+	}
+
+	ca, err := c.getCurrentCABundleContent()
+	if err != nil {
+		return nil, false, nil, err
+	}
+
+	raw, err := sign(ca, x509cr, csr.Spec.Usages, c.certTTL, csr.Spec.ExpirationSeconds, now)
+	if err != nil {
+		return nil, false, nil, err
+	}
+
+	cfg := certificatesv1applyconfigurations.CertificateSigningRequest(name)
+	cfg.Status = certificatesv1applyconfigurations.CertificateSigningRequestStatus().WithCertificate(raw...)
+	return cfg, false, nil, nil
+}
+
+func sign(ca *authority.CertificateAuthority, x509cr *x509.CertificateRequest, usages []certificatesv1.KeyUsage, certTTL time.Duration, expirationSeconds *int32, now func() time.Time) ([]byte, error) {
+	der, err := ca.Sign(x509cr.Raw, authority.PermissiveSigningPolicy{
+		TTL:      duration(certTTL, expirationSeconds),
+		Usages:   usages,
+		Backdate: backdate,       // this must always be less than the minimum TTL requested by a user (see sanity check requestedDuration below)
```

Comment:
Please review backdating & validity - as written, it's the same as upstream k8s signers.